File: fwd_seq_gen.R

#' Calculate the order in which nodes in the tree should be traversed during the forward pass(roots to leaves)
#'
#' Tree is a complex graphical model where we can have multiple parents and multiple children for a node. Hence the order in which the tree should be tranversed becomes significant. Forward algorithm is a dynamic programming problem where to calculate the values at a node,
#' we need the values of the parent nodes beforehand, which need to be traversed before this node. This algorithm outputs a possible(not unique) order of the traversal of nodes ensuring that the parents are traversed first before the children.
#'
#' @param hmm hmm Object of class List given as output by \code{\link{initHMM}}
#' @param nlevel No. of levels in the tree, if known. Default is 100
#' @return Vector of length "D", where "D" is the number of nodes in the tree
#' @examples
#' tmat = matrix(c(0,0,1,0,0,0,0,1,0,0,0,0,0,1,1,0,0,0,0,0,0,0,0,0,0),
#'                5,5, byrow= TRUE ) #for "X" (5 nodes) shaped tree
#' hmmA = initHMM(c("A","B"),list(c("L","R")), tmat) #one feature with two discrete levels "L" and "R"
#' ft_sq = fwd_seq_gen(hmmA)
#' @seealso \code{\link{forward}}

fwd_seq_gen= function(hmm,nlevel=100)
{
  adjm=hmm$adjsym
  pair=which(adjm!=0,arr.ind = TRUE)
  colnames(pair)=NULL
  roots=which(rowSums(adjm)!=0 & colSums(adjm)==0)
  order=list()
  order[[1]]=roots
  for(i in 1:nlevel)
  {
    prev_level=order[[i]]
    nxt_level=c()
    for(j in 1:length(prev_level))
    {
      nxt_nodes=pair[which(pair[,1]==prev_level[j]),2]
      nxt_level=c(nxt_level,nxt_nodes)
    }
    if(length(nxt_level)==0)
    {
      break
    }
    order[[i+1]]=unique(nxt_level)
  }
  order[[(length(order)+1)]]=vector()
  l=length(order)
  for(i in 2:(l-1))
  {
    shift=c()
    for(j in 1:length(order[[i]]))
    {
      from=which(adjm[,order[[i]][j]]!=0)
      fbool=TRUE
      for(m in 1:length(from))
      {
        el=from[m]
        for(k in (i-1):1)
        {
          bool=(el %in% order[[k]])
          if(bool==TRUE)
          {
            break
          }
        }
        fbool= fbool & bool
        if(fbool==FALSE)
        {
          break
        }
      }
      if(fbool==FALSE)
      {
        shift=c(shift,order[[i]][j])
      }
    }
    order[[i]]=unique(order[[i]][! (order[[i]] %in% shift)])
    order[[i+1]]=unique(append(order[[i+1]],shift))
  }
  forder=c()
  for(i in 1:length(order))
  {
    forder=append(forder,order[[i]])
  }
  return(forder)
}

File: src/fwd_seq_gen.py
#!/usr/bin/env python3

# This file calculates the order in which nodes in the tree should be
# traversed during the forward
# pass(indices_of_zero_rowsums_and_nonzero_column_sums to leaves)

# Tree is a complex graphical model where we can have multiple parents and multiple children for a node. Hence the
# order in which the tree should be traversed becomes significant. Forward algorithm is a dynamic programming
# problem where to calculate the values at a node,we need the values of the parent nodes beforehand, which need to be
# traversed before this node.

# Importing the libraries

import numpy as np
import pandas as pd
import initHMM

# Defining the fwd_seq_gen function

def fwd_seq_gen(hmm, number_of_levels=100):
    """
      Args:
          hmm: It is a dictionary given as output by initHMM.py file
          number_of_levels: No. of levels in the tree, if known. Default is 100

      Returns:
          forward_tree_sequence: A list of size "D", where "D" is the number of
          nodes in the tree
      """

    adj_mat = hmm["adjacent_symmetry_matrix"]
    # pair_of_nonzero_indices = np.transpose(np.nonzero(adj_mat))

    # Use this when "adj_mat" is sparse matrix and comment the above line.

    temp = adj_mat.tocoo()
    rows = temp.row
    cols = temp.col
    pair_of_nonzero_indices = np.array([[r,c] for r,c in zip(rows,cols)])

    adj_mat_row_sums = np.squeeze(np.asarray(np.sum(adj_mat, axis=1)))
    adj_mat_col_sums = np.squeeze(np.asarray(np.sum(adj_mat, axis=0)))

    indices_of_zero_rowsums_and_nonzero_column_sums = np.where(
        np.logical_and(
            adj_mat_row_sums != 0,
            adj_mat_col_sums == 0))[0] # np.array()
    order = list()
    order.append(indices_of_zero_rowsums_and_nonzero_column_sums)  # [array]

    for o in order:
        previous_level = o
        next_level = np.array([pair_of_nonzero_indices[list(np.where(pair_of_nonzero_indices[:, 0] == i)[0]), 1] for i in previous_level])
        next_level = np.concatenate(next_level)
        if len(next_level) == 0:
            break
        order.append(np.unique(next_level))
    order.append(np.array([]))
    length_of_order = len(order)

    for i in range(1, length_of_order - 1):
        shift = []
        for j in order[i]:
            indices_of_nonzero_adj_mat_row = np.nonzero(adj_mat[:, j])[0]
            boolean_check = set(indices_of_nonzero_adj_mat_row).issubset(
                np.hstack(order[:i]))
            if not boolean_check:
                shift.append(j)
        element_to_update = [i for i in order[i] if i not in shift]
        order[i] = np.unique(element_to_update)
        order[i + 1] = np.unique(list(order[i + 1]) + shift)
    forward_order = []
    for i in order:
        forward_order = forward_order + list(i)

    return forward_order


def run_an_example():
    """
    sample run for fwd_seq_gen function
    """
    import initHMM
    sample_tree = np.array([0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
                            1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]).reshape(5, 5)
    states = ['P', 'N']
    symbols = [['L', 'R']]
    hmm = initHMM.initHMM(states, symbols, sample_tree)

    forward_tree_sequence = fwd_seq_gen.fwd_seq_gen(hmm)

if __name__ == "__main__":
    run_an_example()
